sympy/functions/elementary/piecewise.py

```diff
@@ -227,242 +330,468 @@ def _eval_derivative(self, x):
-                        newargs.append((values[i], (c == True and i == len(values) - 1) or
-                            And(rep >= intervals[i][0], rep <= intervals[i][1])))
-            return self.func(*newargs)
+            return super(Piecewise, self)._eval_interval(sym, a, b)
+        else:
+            x, lo, hi = map(as_Basic, (sym, a, b))
 
-        # Determine what intervals the expr,cond pairs affect.
-        int_expr = self._sort_expr_cond(sym, a, b)
+        if _first:  # get only x-dependent relationals
+            def handler(ipw):
+                if isinstance(ipw, self.func):
+                    return ipw._eval_interval(x, lo, hi, _first=None)
+                else:
+                    return ipw._eval_interval(x, lo, hi)
+            irv = self._handle_irel(x, handler)
+            if irv is not None:
+                return irv
+
+            if (lo < hi) is S.false or (
+                    lo is S.Infinity or hi is S.NegativeInfinity):
+                rv = self._eval_interval(x, hi, lo, _first=False)
+                if isinstance(rv, Piecewise):
+                    rv = Piecewise(*[(-e, c) for e, c in rv.args])
+                else:
+                    rv = -rv
+                return rv
 
-        # Finally run through the intervals and sum the evaluation.
-        ret_fun = 0
-        for int_a, int_b, expr in int_expr:
-            if isinstance(expr, Piecewise):
-                # If we still have a Piecewise by now, _sort_expr_cond would
-                # already have determined that its conditions are independent
-                # of the integration variable, thus we just use substitution.
-                ret_fun += piecewise_fold(
-                    super(Piecewise, expr)._eval_interval(sym, Max(a, int_a), Min(b, int_b)))
+            if (lo < hi) is S.true or (
+                    hi is S.Infinity or lo is S.NegativeInfinity):
+                pass
             else:
-                ret_fun += expr._eval_interval(sym, Max(a, int_a), Min(b, int_b))
-        return mul * ret_fun
-
-    def _sort_expr_cond(self, sym, a, b, targetcond=None):
-        """Determine what intervals the expr, cond pairs affect.
-
-        1) If cond is True, then log it as default
-        1.1) Currently if cond can't be evaluated, throw NotImplementedError.
-        2) For each inequality, if previous cond defines part of the interval
-           update the new conds interval.
-           -  eg x < 1, x < 3 -> [oo,1],[1,3] instead of [oo,1],[oo,3]
-        3) Sort the intervals to make it easier to find correct exprs
-
-        Under normal use, we return the expr,cond pairs in increasing order
-        along the real axis corresponding to the symbol sym.  If targetcond
-        is given, we return a list of (lowerbound, upperbound) pairs for
-        this condition."""
+                a = Dummy('lo')
+                b = Dummy('hi')
+                pos = self._eval_interval(x, a, b, _first=False)
```

Comment:
Does this look like the right thing to do with Numeric/non-Numeric args?

Also, instead of returning a Piecewise, perhaps a single expression with `sign(hi - lo)*pos` should be returned.

Reply:
Hi, I agree that there is a problem with Piecewise itself. It seems nonstandard, and maybe creating unnecessary implementation difficulties? Shouldn't a Piecewise function be just a list of intervals with open boundaries on both sides? And the syntax should require both the upper and lower bound on each and every interval? 
I don't think it unreasonable to expect the user to be completely explicit in defining their input function in this way. 
Or would that entail changes too widespread throughout sympy?

Reply:
I cannot think of a better way for non-numeric args.

The return value should probably be Piecewise. I'm not sure that the single expression would be correct (unless it is Piecewise in disguise: `(sign + 1)*pos/2 + (-sign + 1)*neg/2`).

Reply:
I don't know if it's obvious to anyone but changing the order of the limits does not only change the sign of the result, it involves a swap of the symbols representing lower and upper limits. Here is a piece of code taken from the tests added that demonstrates the inequality for a fairly simple expression:

```python
>>> from sympy.utilities.iterables import *
>>> lim = Tuple(x, c, d)
>>> p = Piecewise((1, x < a), (2, x > b), (3, True))
>>> q = p.integrate(lim)
>>> assert q == Piecewise(
...     (-c + 2*d - 2*Min(d, Max(a, c)) + Min(d, Max(a, b, c)), c < d),
...     (-2*c + d + 2*Min(c, Max(a, d)) - Min(c, Max(a, b, d)), True))
>>> pos = q.args[0].expr
>>> eq=Eq(q, sign(d-c)*pos)  # is the general result equivalent to a signed pos value?
>>> for v in permutations((1, 2, 3, 4)):
...     r = dict(zip((a, b, c, d), v))
...     if eq.subs(r) is not S.true:
...         print r
...         break
{c: 3, b: 2, a: 1, d: 4}  # answer: no. These values demonstrate a counter-point.
```

Reply:
(In the test suite I exhaustively test the original expression with substituted values and then integrate with the known limits and compare it against the general result.)

Reply:
#13552 turns out to be a good test for piecewise integration.
 
from sympy.stats import *
mixed = Uniform("u1",-1,1) + Uniform("u2",-1,1)
density(mixed)

It seems that uniform densities in sympy.stats are written in terms of Heaviside functions instead of Piecewise expressions. That leads to a rather intractable expression for the density function that consists of two linear pieces: first is increases linearly from 0 to 1/2 in the interval (-2, 0) and then decreases linearly from 1/2 to 0 in the interval (0, 2).

Reply:
```
>>> p=Piecewise((0,x<-2),(x/4+S.Half,x<=0),(-x/4+S.Half,x<=2),(0,True))
>>> [p.subs(x,i) for i in range(-3,4)]
[0, 0, 1/4, 1/2, 1/4, 0, 0]
>>> p.integrate(x)
Piecewise((0, x < -2), (x**2/8 + x/2 + 1/2, x <= 0), (-x**2/8 + x/2 + 1/2, x <=2), (1, True))
```

I think I have a step or similar function in the test suite. I'll double check.

Abs.rewrite(Piecewise) should also be defined. The equivalent expression written in terms of Abs is unevaluated:

```
>>> Piecewise((S.Half - abs(x/4), (x>=-2)&(x<=2)), (0,True)).integrate(x)
Piecewise((0, x <= -2), (-(Integral(-2, x) + Integral(Abs(x), x))/4 + Integral(-
2, (x, -2))/4 + Integral(Abs(x), (x, -2))/4, x <= 2), (Integral(-2, (x, -2))/4 -
 Integral(-2, (x, 2))/4 + Integral(Abs(x), (x, -2))/4 - Integral(Abs(x), (x, 2))/4, True))

>>> _.subs(Abs(x),Piecewise((x,x>=0),(-x,True)))
Piecewise((-Piecewise((x, x >= 0), (-x, True))/4 + 1/2, (x >= -2) & (x <= 2)), (0, True))
>>> _.integrate(x)
Piecewise((0, x <= -2), (x**2/8 + x/2 + 1/2, x <= 0), (-x**2/8 + x/2 + 1/2, x <= 2), (1, True))
```

Reply:
The triangle-wave shape is integrated in a test at line 205:

```
    g = Piecewise((1 - x, Interval(0, 1).contains(x)),
        (1 + x, Interval(-1, 0).contains(x)), (0, True))
    assert integrate(g, (x, -5, 1)) == 1
    assert integrate(g, (x, 1, -5)) == -1
    assert integrate(g, (x, 1, y)).subs(y, -5) == -1
    assert integrate(g, (x, y, -5)).subs(y, 1) == -1
    i = integrate(g, (x, -5, y))
    assert i.subs(y, 1) == 1
    assert piecewise_fold(i.rewrite(Piecewise)
        ) == Piecewise(
        (1, y >= 1),
        (-y**2/2 + y + 1/2, y >= 0),
        (y**2/2 + y + 1/2, y >= -1),
        (0, True))
    assert i == Piecewise(
        (-Min(-1, y)**2/2 - Min(-1, y) +
            Min(0, y)**2 - Min(1, y)**2/2 + Min(1, y), y > -5),
        (0, True))
    i = integrate(g, (x, y, 1))
    assert i.subs(y, -5) == 1
    assert piecewise_fold(i.rewrite(Piecewise)
        )== Piecewise(
        (1, y <= -1),
        (-y**2/2 - y + 1/2, y <= 0),
        (y**2/2 - y + 1/2, y < 1),
        (0, True))
    assert i == Piecewise(
        (Min(1, Max(0, y))**2 - Min(1, Max(-1, y), Max(0, y))**2/2 -
            Min(1, Max(-1, y), Max(0, y)) + 1/2, y < 1),
        (0, True))
```

This function is defined in a variety of ways (motivated by the original issue). I'm not sure that all those ways as needed...those are slow tests. Perhaps it would be better to leave the variety of conditions to be tested with simpler functions (e.g. `Piecewise((1, cond1),(2, cond2), (0, True))`.